image_segmentation/pytorch/runtime/trainer/xla_trainer.py

```diff
@@ -46,6 +49,42 @@ def train_step(
         self, iteration: int, images: torch.Tensor, labels: torch.Tensor
     ) -> torch.Tensor:
         """Overrides UNet3DTrainer.train_step"""
+        trace_context = (
+            xp.Trace("build_graph")
+            if hasattr(self, "profile_server")
+            else contextlib.nullcontext()
+        )
```

Comment:
Should `xp.Trace` logic be kept within `xla_trainer` in the first place?

E.g. in `UNet3DRunner` we should have functions such as `get_trace_context()` which returns `nullcontext` by default for the non XLA runners and the logic currently in `UNet3DRunner` can be kept in `XLATrainer`. This way the base class doesn't need to import `torch_xla.debug.profiler` at all which we shouldn't do in non XLA scripts.

Reply:
Good point. We should not expose any `torch_xla` logic in UNet3DTrainer as it is meant to be abstract.

I have moved all `torch_xla.debug.profiler` code to inside `XLATrainer` and changed `UNet3DTrainer` to have an abstract `get_step_trace_context` method.

I also added additional smoke tests for the new `XLATrainer.get_step_trace_context`.

Reply:
Thanks!